docs/man/man1/axis.1
.\" Copyright (c) 2007 Alex Joni
.\"                (alex_joni AT users DOT sourceforge DOT net)
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
.\" USA.
.\"
.\"
.\"
.TH AXIS "1"  "2007-04-01" "LinuxCNC Documentation" "The Enhanced Machine Controller"
.SH NAME
axis \- AXIS LinuxCNC Graphical User Interface
.SH SYNOPSIS
.B axis
\fI\-ini\fR \fIINIFILE\fR
.SH DESCRIPTION
\fBaxis\fR is one of the Graphical User Interfaces (GUI) for LinuxCNC.
It gets run by the runscript usually.
.SH OPTIONS
.TP
\fBINIFILE\fR
The INI file is the main piece of an LinuxCNC configuration.
It is not the entire configuration; there are various other files that go with it
(NML files, HAL files, TBL files, VAR files). It is, however, the most important one,
because it is the file that holds the configuration together.
It can adjust a lot of parameters itself, but it also tells \fBLinuxCNC\fR which other files to load and use.

.SH "SEE ALSO"
\fBLinuxCNC(1)\fR

Much more information about LinuxCNC and HAL is available in the LinuxCNC
and HAL User Manuals, found at /usr/share/doc/LinuxCNC/.

.SH BUGS
None known at this time. 
.PP
.SH AUTHOR
This man page written by Alex Joni, as part of the LinuxCNC project.
.SH REPORTING BUGS
Report bugs at https://github.com/LinuxCNC/linuxcnc/issues
.SH COPYRIGHT
Copyright \(co 2007 Alex Joni.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


docs/NEWS
check debian/changelog for recent news

# some of EMC2 History & News
#############################

2009.06.13 - mshaver added emc2/configs/smithy directory and related files
             for Smithy Company's line of cnc machines.

2006.11.17 - removed HAL types s8, u8, s16, and u16.  HAL now uses float
	     for analog values, bit for booleans, and s32 or u32 for
	     integer values.  The only users who will be affected are
	     those with Universal Stepper Controller cards.  See item 9 at
	     http://wiki.linuxcnc.org/cgi-bin/emcinfo.pl?UpdatingConfigurationsForDevelopmentVersions

2006.11.15 - merged classicladder 0.7.100

2006.11.04 - added a feed-hold HAL pin.  Added M codes to enable/disable
	     feed override, spindle override, feed hold, and adaptive feed
	     individually.  Feed override is M50, spindle override is M51,
	     adaptive feed is M52, and feed hold is M53.  G50/51 are no
	     longer used to enable/disable adaptive feed, use M52 instead.

2006.10.24 - added support for block delete as defined by Section 2.2.2
             in the RS274NGC_3 specs.

2006.10.24 - added support for optional program stop as defined by Section 
             2.2.3 in the RS274NGC_3 specs.

2006.10.14 - manpages have been created for most rtapi and hal calls

2006.10.01 - added a userspace simulator which includes rtapi threads.
	     Use "--enable-simulator" to enable this mode, which cannot
	     be used to control real hardware

2006.09.27 - added support to optionally save joint positions between
             runs ([TRAJ] POSITION_FILE = position.txt)

2006.09.26 - added support to specify UNITS in literal in the ini. 
             (e.g. [TRAJ] LINEAR_UNITS = mm )

2006.09.09 - fixed bug #1384883, optional stop was not optional.

2006.08.09 - AXIS is now a part of EMC2, not a separate source download.

2006.08.07 - Added a spindle-speed override. This allows to override the spindle
	     speed during a program run. The MIN_SPINDLE_OVERRIDE and 
	     MAX_SPINDLE_OVERRIDE in the ini assure that speed doesn't get 
	     reduced or increased too much. Both values are in %.

2006.08.06 - Documentation can be built from source with the 'make docs' target

2006.06.11 - added realtime support for jogwheels.  Three HAL pins per axis:
	     one accepts raw counts from the wheel, one sets the distance moved
	     per count, and one enables that axis to move.

2006.05.17 - added 'adaptive feed', a HAL pin that can be used to scale the
	     feedrate in real time.  Simular to feedrate override, but can
	     only vary from 0.0 to 1.0, and reacts much faster.  Intended 
	     for use with EDM and other applications that require closed 
	     loop control of feedrate.

2006.03.18 - A new pin "count" exists for freqgen.  It is the same as the
             internal symbol "count" shown in the hal introduction.pdf,
             or the result of up OR down if there was an "OR" component.

2006.03.16 - Modified the motion control module to behave more like other
             HAL modules - it no longer adds its functions to threads
             automatically.  All the sample configs have been changed to
             explicitly add the functions, users who have custom configs
             will need to do the same.

2006.03.12 - If a copy of axis 1.3a0 or newer is detected inside src/
             it will be automatically built and installed for
             --enable-run-in-place use

2006.03.10 - minor bugfix in tkemc for jogging

2006.03.10 - added genhexkins, with a minitetra sample config.
             This proves EMC2 works just as good with hexapods as EMC1 did.

2006.03.10 - added VTI support & sample configs

2006.03.07 - lots of fixes to the TP (most were wrong in the old one as well)

2006.03.05 - new Trajectory Planner

2006.02.20 - added a manpage for the runscript.

2006.02.12 - Fixed a bug that was causing the RT code to run very slowly,
             forcing users to set BASE_PERIOD very long.

2006.02.10 - The build process has been replaced, no more recursive $(MAKE)

2006.01.28 - fixed bug #1205237, mode switching sometimes hung

2006.01.08 - M62/M63 digital outputs synchronous to motion
             M64/M65 digital outputs that get set right away

2006.01.06 - work on halconfig.tcl, it is used to display/configure HAL data
             requires the BWidget package, realtime, and HAL to run.

2006.01.04 - fix bug #1387740, a motion that goes nowhere clears feedrate

2006.01.01 - the config picker has i18n support, a German translation exists

2005.12.20 - a new config picker has been written, if you start just the
             runscript this one helps you choose a config to run

2005.12.13 - new runscript, and config organization
             work towards the first release of EMC2

2005.xx.xx - EMC2 now supports tool-changing in complex scenarios

2005.xx.xx - classicladder now a HAL component, a software PLC

2004.xx.xx - HAL (the Hardware Abstraction Layer) is used for
	     addressing hardware drivers, and allows for simple configuration


#developers: this file should hold a history of improvements, bugfixes, etc
             it is useful for keeping track of development


src/emc/usr_intf/stepconf/axis.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#    This is stepconf, a graphical configuration editor for LinuxCNC
#    Copyright 2007 Jeff Epler <jepler@unpythonic.net>
#    stepconf 1.1 revamped by Chris Morley 2014
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

#********************
# AXIS GENERAL PAGE
#********************
import sys
import time
import hal
from gi.repository import Gtk
from gi.repository import GObject
from stepconf import preset
from stepconf.definitions import *

def axis_prepare(self, axis):
	def set_text(n):
		self.w[axis + n].set_text("%s" % self.d[axis + n])
		# Set a name for this widget. Necessary for css id
		self.w[axis + n].set_name("%s%s" % (axis, n))
	def set_active(n):
		self.w[axis + n].set_active(self.d[axis + n])

	set_text("steprev")
	set_text("microstep")
	set_text("pulleynum")
	set_text("pulleyden")
	set_text("leadscrew")
	set_text("maxvel")
	set_text("maxacc")
	set_text("homepos")
	set_text("minlim")
	set_text("maxlim")
	set_text("homesw")
	set_text("homevel")
	set_active("latchdir")

	if axis == "a":
		self.w[axis + "screwunits"].set_text(_("degree / rev"))
		self.w[axis + "velunits"].set_text(_("deg / s"))
		self.w[axis + "accunits"].set_text(_(u"deg / s²"))
		self.w[axis + "accdistunits"].set_text(_("deg"))
		self.w[axis + "scaleunits"].set_text(_("Steps / deg"))
	elif self.d.units:
		self.w[axis + "screwunits"].set_text(_("mm / rev"))
		self.w[axis + "velunits"].set_text(_("mm / s"))
		self.w[axis + "accunits"].set_text(_(u"mm / s²"))
		self.w[axis + "accdistunits"].set_text(_("mm"))
		self.w[axis + "scaleunits"].set_text(_("Steps / mm"))
	else:
		self.w[axis + "screwunits"].set_text(_("rev / in"))
		self.w[axis + "velunits"].set_text(_("in / s"))
		self.w[axis + "accunits"].set_text(_(u"in / s²"))
		self.w[axis + "accdistunits"].set_text(_("in"))
		self.w[axis + "scaleunits"].set_text(_("Steps / in"))

	#self.check_switch_limits(axis)

	self.w[axis + "steprev"].grab_focus()
	GObject.idle_add(lambda: self.update_pps(axis))
	
	# Prepare preset
	preset_index = self.d[axis + "preset"]
	self.d.select_combo_machine(self.w[axis + "preset_combo"], preset_index)
	self.preset_axis(axis)

def axis_done(self, axis):
	def get_text(n):
		self.d[axis + n] = float(self.w[axis + n].get_text())
	def get_active(n):
		self.d[axis + n] = self.w[axis + n].get_active()
	get_text("steprev")
	get_text("microstep")
	get_text("pulleynum") 
	get_text("pulleyden")
	get_text("leadscrew")
	get_text("maxvel")
	get_text("maxacc")
	get_text("homepos")
	get_text("minlim")
	get_text("maxlim")
	get_text("homesw")
	get_text("homevel")
	get_active("latchdir")
	# Save preset
	current_machine = self.d.get_machine_preset(self.w[axis + "preset_combo"])
	if current_machine:
		self.d[axis + "preset"] = current_machine["index"]
	else:
		# Other selected
		self.d[axis + "preset"] = 0

def check_switch_limits(self, axis):
	inputs = self.a.build_input_set()
	#thisaxishome = set([ALL_HOME, ALL_LIMIT_HOME, "home-" + axis, "min-home-" + axis,
	#					"max-home-" + axis, "both-home-" + axis])
	thisaxishome = set([d_hal_input[ALL_HOME], d_hal_input[ALL_LIMIT_HOME], "home-" + axis, "min-home-" + axis,
						"max-home-" + axis, "both-home-" + axis])
	# Test if exists limit switches
	homes = bool(inputs & thisaxishome)
	if (homes == False):
		self.w[axis + "homesw"].set_sensitive(homes)
		self.w[axis + "homevel"].set_sensitive(homes)
		self.w[axis + "latchdir"].set_sensitive(homes)
	else:
		current_machine = self.d.get_machine_preset(self.w[axis + "preset_combo"])
		if current_machine:
			if (axis + "homesw" in current_machine):
				self.w[axis + "homesw"].set_sensitive(0)
			if (axis + "homevel" in current_machine):
				self.w[axis + "homevel"].set_sensitive(0)
			if (axis + "latchdir" in current_machine):
				self.w[axis + "latchdir"].set_sensitive(0)
		else:
			# Other selected
			self.w[axis + "homesw"].set_sensitive(homes)
			self.w[axis + "homevel"].set_sensitive(homes)
			self.w[axis + "latchdir"].set_sensitive(homes)
			return

def preset_axis(self, axis):
	# List axis page widgets (except latchdir)
	lwidget=[
		axis + "steprev",
		axis + "microstep",
		axis + "pulleynum",
		axis + "pulleyden",
		axis + "leadscrew",
		axis + "maxvel",
		axis + "maxacc",
		axis + "homepos",
		axis + "minlim",
		axis + "maxlim",
		axis + "homesw",
		axis + "homevel"
	]

	current_machine = self.d.get_machine_preset(self.w[axis + "preset_combo"])
	if current_machine:
		None
	else:
		# Other selected
		for w in lwidget:
			self.w['%s'%w].set_sensitive(1)
		self.check_switch_limits(axis)
		return

	for w in lwidget:
		if(w in current_machine):
			self.w['%s'%w].set_text(str(current_machine[w]))
			self.w['%s'%w].set_sensitive(0)
		else:
			self.w['%s'%w].set_sensitive(1)
	if(axis + "latchdir" in current_machine):
		self.w['%slatchdir'%axis].set_active(current_machine[axis + "latchdir"])
		#self.w['%slatchdir'%axis].set_sensitive(0)
	#else:
	#	self.w['%slatchdir'%axis].set_sensitive(1)
	# 
	self.check_switch_limits(axis)


# for Axis page calculation updates
def update_pps(self, axis):
	def get(n):
		return float(self.w[axis + n].get_text())
	self.axis_sanity_test(axis)
	try:
		pitch = get("leadscrew")
		step = get("steprev")
		micro = get("microstep")
		pullnum = get("pulleynum")
		pulldem = get("pulleyden")
		if self.d.units == 1 or axis == 'a':
			pitch = 1./pitch
		pps = (pitch * step * micro * (pullnum / pulldem) * get("maxvel"))
		if pps == 0:
			raise ValueError
		pps = abs(pps)
		acctime = get("maxvel") / get("maxacc")
		accdist = acctime * .5 * get("maxvel")
		self.w[axis + "acctime"].set_text("%.4f" % acctime)
		self.w[axis + "accdist"].set_text("%.4f" % accdist)
		self.w[axis + "hz"].set_text("%.1f" % pps)
		scale = self.d[axis + "scale"] = (1.0 * pitch * step * micro * (pullnum / pulldem))
		self.w[axis + "scale"].set_text("%.1f" % scale)
		temp = "Axis Scale: %d × %d × (%.1f ÷ %.1f) × %.3f ="% (int(step),int(micro),(pullnum),(pulldem),pitch)
		self.w[axis + "scaledescr"].set_text(temp)
		self.set_buttons_sensitive(1,1)
		self.w[axis + "axistest"].set_sensitive(1)
	except (ValueError, ZeroDivisionError): # Some entries not numbers or not valid
		self.w[axis + "acctime"].set_text("")
		self.w[axis + "accdist"].set_text("")
		self.w[axis + "hz"].set_text("")
		self.w[axis + "scale"].set_text("")
		self.set_buttons_sensitive(0,0)
		self.w[axis + "axistest"].set_sensitive(0)

def axis_sanity_test(self, axis):
	# I hate the inner function
	def get(n):
		return float(self.w[axis + n].get_text())

	# List of field with background color can change
	datalist = ('steprev','microstep','pulleynum','pulleyden','leadscrew',
				'maxvel','maxacc')
	mystyle =""
	for i in datalist:
		# Damn! this is a bug. GTKBuilder sets the widget name to be the builder ID.
		#widget_name = Gtk.Buildable.get_name(self.w[axis+i])
		ctx = self.w[axis+i].get_style_context()
		try:
			a=get(i)
			if a <= 0:
				raise ValueError
		except:
			ctx.add_class('invalid')
		else:
			ctx.remove_class('invalid')

#**********
# Axis Test
#***********
def test_axis(self, axis):
	if not self.check_for_rt():
		return

	# Retrive user setting for maxvel and maxacc on current axis tab
	# Test if there is some data saved
	testvel = self.d[axis + "testmaxvel"]
	testacc = self.d[axis + "testmaxacc"]
	# Check if not null and not empty and not string "None"
	if(testvel != None and testvel != "" and testvel != "None"):
		vel = float(testvel)
	else:
		vel = float(self.w[axis + "maxvel"].get_text())

	# Check if not null and not empty and not string "None"
	if(testacc != None and testacc != "" and testacc != "None"):
		acc = float(testacc)
	else:
		acc = float(self.w[axis + "maxacc"].get_text())

	scale = self.d[axis + "scale"]
	maxvel = float(self.w[axis + "maxvel"].get_text()) * 1.5
	#maxvel = 1.5 * vel
	if self.a.doublestep():
			period = int(1e9 / maxvel / scale)
	else:
			period = int(.5e9 / maxvel / scale)
	
	steptime = self.w.steptime.get_value()
	stepspace = self.w.stepspace.get_value()
	latency = self.w.latency.get_value()
	minperiod = self.d.minperiod()
	
	if period < minperiod:
		period = minperiod
		if self.a.doublestep():
			maxvel = 1e9 / minperiod / abs(scale)
		else:
			maxvel = 1e9 / minperiod / abs(scale)
	if period > 100000:
		period = 100000

	# halrun print a point "." on console, but I have not found out why.
	self.halrun = os.popen("halrun -Is", "w")
	if self.a.debug:
		self.halrun.write("echo\n")
	axnum = "xyza".index(axis)
	step = axis + "step"
	dir = axis + "dir"
	
	self.halrun.write("""
		loadrt steptest
		loadrt stepgen step_type=0
		""")

	port3name=port2name=port2dir=port3dir=""
	if self.d.number_pports>2:
		 port3name = ' '+self.d.ioaddr3
		 if self.d.pp3_direction: # Input option
			port3dir =" in"
		 else: 
			port3dir =" out"
	if self.d.number_pports>1:
		 port2name = ' '+self.d.ioaddr2
		 if self.d.pp2_direction: # Input option
			port2dir =" in"
		 else: 
			port2dir =" out"
	self.halrun.write( "loadrt hal_parport cfg=\"%s out%s%s%s%s\"\n" % (self.d.ioaddr, port2name, port2dir, port3name, port3dir))
	self.halrun.write("""
		loadrt threads period1=%(period)d name1=fast fp1=0 period2=1000000 name2=slow
		addf stepgen.make-pulses fast
		addf parport.0.write fast
		"""%{'period': period})
	
	if self.d.number_pports>1:
		self.halrun.write( "addf parport.0.write fast\n")
	if self.d.number_pports>2:
		self.halrun.write( "addf parport.0.write fast\n")
	temp = self.a.find_output(axis +'step')
	step_pin = temp[0][0]
	temp = self.a.find_output(axis +'dir')
	dir_pin = temp[0][0]
	self.halrun.write("""
		addf stepgen.capture-position slow
		addf steptest.0 slow
		addf stepgen.update-freq slow

		net step stepgen.0.step => parport.0.pin-%(steppin)02d-out
		net dir stepgen.0.dir => parport.0.pin-%(dirpin)02d-out
		net cmd steptest.0.position-cmd => stepgen.0.position-cmd
		net fb stepgen.0.position-fb => steptest.0.position-fb

		setp parport.0.pin-%(steppin)02d-out-reset 1
		setp stepgen.0.steplen 1
		setp stepgen.0.dirhold %(dirhold)d
		setp stepgen.0.dirsetup %(dirsetup)d
		setp stepgen.0.position-scale %(scale)f
		setp steptest.0.epsilon %(onestep)f

		setp stepgen.0.enable 1
	""" % {
		'steppin': step_pin,
		'dirpin': dir_pin,
		'dirhold': self.d.dirhold + self.d.latency,
		'dirsetup': self.d.dirsetup + self.d.latency,
		'onestep': abs(1. / self.d[axis + "scale"]),
		'scale': self.d[axis + "scale"],
	})
	
	if self.a.doublestep():
		self.halrun.write("""
			setp parport.0.reset-time %(resettime)d
			setp stepgen.0.stepspace 0
			addf parport.0.reset fast
		""" % {
			'resettime': self.d['steptime']
		})
	amp_signals = self.a.find_output(d_hal_output[AMP])
	for pin in amp_signals:
		amp,amp_port = pin
		self.halrun.write("setp parport.%(portnum)d.pin-%(enablepin)02d-out 1\n"
			% {'enablepin': amp,'portnum': amp_port})
	
	estop_signals = self.a.find_output(d_hal_output[ESTOP])
	for pin in estop_signals:
		estop,e_port = pin
		self.halrun.write("setp parport.%(portnum)d.pin-%(estoppin)02d-out 1\n"
			% {'estoppin': estop,'portnum': e_port})
	
	for pin in 1,2,3,4,5,6,7,8,9,14,16,17:
		inv = getattr(self.d, "pin%dinv" % pin)
		if inv:
			self.halrun.write("setp parport.0.pin-%(pin)02d-out-invert 1\n"
				% {'pin': pin})
	if self.d.number_pports > 1:
		if self.d.pp2_direction:# Input option
			out_list =(1,14,16,17)
		else:
			out_list =(1,2,3,4,5,6,7,8,9,14,16,17)
		for pin in (out_list):
			inv = getattr(self.d, "pp2_pin%dinv" % pin)
			if inv:
				self.halrun.write("setp parport.1.pin-%(pin)02d-out-invert 1\n"
				% {'pin': pin})
	if self.a.debug:
		self.halrun.write("loadusr halmeter sig cmd -g 275 415\n")
	
	if axis == "a":
		self.w.testvelunit.set_text(_("deg / s"))
		self.w.testaccunit.set_text(_(u"deg / s²"))
		self.w.testampunit.set_text(_("deg"))
		self.w.testvel.set_increments(1,5)
		self.w.testacc.set_increments(1,5)
		self.w.testamplitude.set_increments(1,5)
		self.w.testvel.set_range(0, maxvel)
		self.w.testacc.set_range(1, 360000)
		self.w.testamplitude.set_range(0, 1440)
		self.w.testvel.set_digits(1)
		self.w.testacc.set_digits(1)
		self.w.testamplitude.set_digits(1)
		self.w.testamplitude.set_value(10)
	elif (self.d.units == MM):
		self.w.testvelunit.set_text(_("mm / s"))
		self.w.testaccunit.set_text(_(u"mm / s²"))
		self.w.testampunit.set_text(_("mm"))
		self.w.testvel.set_increments(1,5)
		self.w.testacc.set_increments(1,5)
		self.w.testamplitude.set_increments(1,5)
		self.w.testvel.set_range(0, maxvel)
		self.w.testacc.set_range(1, 100000)
		self.w.testamplitude.set_range(0, 1000)
		self.w.testvel.set_digits(2)
		self.w.testacc.set_digits(2)
		self.w.testamplitude.set_digits(2)
		self.w.testamplitude.set_value(15)
	else:
		self.w.testvelunit.set_text(_("in / s"))
		self.w.testaccunit.set_text(_(u"in / s²"))
		self.w.testampunit.set_text(_("in"))
		self.w.testvel.set_increments(.1,5)
		self.w.testacc.set_increments(1,5)
		self.w.testamplitude.set_increments(.1,5)
		self.w.testvel.set_range(0, maxvel)
		self.w.testacc.set_range(1, 3600)
		self.w.testamplitude.set_range(0, 36)
		self.w.testvel.set_digits(1)
		self.w.testacc.set_digits(1)
		self.w.testamplitude.set_digits(1)
		self.w.testamplitude.set_value(.5)
	
	self.jogplus = self.jogminus = 0
	self.w.testdir.set_active(0)
	self.w.run.set_active(0)
	self.w.testacc.set_value(acc)
	self.w.testvel.set_value(vel)
	self.a.axis_under_test = axis
	self.update_axis_test()

	self.halrun.write("start\n")
	self.halrun.flush()

	self.w.dlgTestAxis.set_title(_("%s Axis Test") % axis.upper())
	self.w.dlgTestAxis.show_all()
	result = self.w.dlgTestAxis.run()
	if result == Gtk.ResponseType.OK:
		# Save test parameters
		self.d[axis + "testmaxvel"] = self.w.testvel.get_value()
		self.d[axis + "testmaxacc"] = self.w.testacc.get_value()
	self.w.dlgTestAxis.hide()
	
	if amp_signals:
		for pin in amp_signals:
			amp,amp_port = pin
			self.halrun.write("setp parport.%(portnum)d.pin-%(enablepin)02d-out 0\n"
			% {'enablepin': amp,'portnum': amp_port})
	if estop_signals:
		for pin in estop_signals:
			estop,e_port = pin
			self.halrun.write("setp parport.%(portnum)d.pin-%(estoppin)02d-out 0\n"
			% {'estoppin': estop,'portnum': e_port})

	time.sleep(.001)
	self.halrun.close()


# Check for realtime-capable LinuxCNC.
# Returns True if the running version of LinuxCNC is realtime-capable
# (or if debug is enabled), returns False otherwise.
def check_for_rt(self):
	is_realtime_capable = False
	try:
		if hal.is_sim:
			self.a.warning_dialog(MESS_NO_REALTIME,True)
		elif hal.is_rt:
			if hal.is_kernelspace:
				actual_kernel = os.uname()[2]
				if hal.kernel_version == actual_kernel:
					is_realtime_capable = True
				else:
					self.a.warning_dialog(MESS_KERNEL_WRONG + '%s'%hal.kernel_version,True)
			else:
				is_realtime_capable = True
	except:
		print 'STEPCONF WARNING: check-for-realtime function failed - continuing anyways.'
		print sys.exc_info()
		return True

	if is_realtime_capable or self.a.debug:
		return True
	else:
		return False

#*********************
# General Axis methods and callbacks
#*********************
def update_axis_params(self, *args):
	self.update_axis_test()

def update_axis_test(self):
	axis = self.a.axis_under_test
	if axis is None:
		return

	self.halrun.write("""
		setp stepgen.0.maxaccel %(accel)f
		setp stepgen.0.maxvel %(vel)f
		setp steptest.0.jog-minus %(jogminus)s
		setp steptest.0.jog-plus %(jogplus)s
		setp steptest.0.run %(run)s
		setp steptest.0.amplitude %(amplitude)f
		setp steptest.0.maxvel %(vel)f
		setp steptest.0.dir %(dir)s
	""" % {
		'jogminus': self.a.jogminus,
		'jogplus': self.a.jogplus,
		'run': self.w.run.get_active(),
		'amplitude': self.w.testamplitude.get_value(),
		'accel': self.w.testacc.get_value(),
		'vel': self.w.testvel.get_value(),
		'dir': self.w.testdir.get_active(),
	})
	self.halrun.flush()

def on_jogminus_pressed(self, *args):
	self.a.jogminus = 1
	self.update_axis_test()

def on_jogminus_released(self, *args):
	self.a.jogminus = 0
	self.update_axis_test()

def on_jogplus_pressed(self, *args):
	self.a.jogplus = 1
	self.update_axis_test()

def on_jogplus_released(self, *args):
	self.a.jogplus = 0
	self.update_axis_test()
